README.md
VersionOne SDK - A Ruby SDK for the VersionOne REST API
=======================================================

[![Gem Version][gem-version-svg]][gem-version-link]
[![Build Status][build-status-svg]][build-status-link]
[![Coverage Status][coverage-status-svg]][coverage-status-link]
[![Dependency Status][dependency-status-svg]][dependency-status-link]
[![Code Climate][codeclimate-status-svg]][codeclimate-status-link]
[![Scrutinizer Code Quality][scrutinizer-status-svg]][scrutinizer-status-link]
[![Downloads][downloads-svg]][downloads-link]
[![Docs][docs-rubydoc-svg]][docs-rubydoc-link]
[![License][license-svg]][license-link]

## Synopsis

This is a VersionOne SDK in Ruby that accesses the VersionOne REST API.

VersionOne currently offers SDKs in Java, .NET, Python and JavaScript but not in Ruby or other languages. This SDK seeks to provide an easier way to use the REST API for Ruby applications.

It currently offers the following capabilities:

1. Ability to retrieve and parse all Assets of a certain type to JSON (via JsonDoc)
2. Ability to query Assets transparently using Asset OID Tokens (e.g. `Story:1`) or Asset Numbers (e.g. `B-1`).
3. Ability to update Assets using Ruby without needing to manually create XML.

Note: This SDK doesn't currently support NTLM authentication used with on-premise VersionOne deployments. An approach for this is to use [`httpclient`](https://rubygems.org/gems/httpclient) and [`rubyntlm`](https://rubygems.org/gems/rubyntlm).

## Installation

### Via Bundler

Add 'versionone_sdk' to `Gemfile` and then run `bundle`:

```sh
$ echo "gem 'versionone_sdk'" >> Gemfile
$ bundle
```

### Via RubyGems

```sh
$ gem install versionone_sdk
```

This gem uses `nokogiri` which requires Ruby >= 1.9.2.

## Examples

```ruby
require 'versionone_sdk'

# Authorizing using Basic Authentication
params = {
  hostname: 'www1.v1host.com',
  instance: 'myinstance',
  username: 'myusername',
  password: 'mypassword',
  port: 443,
  protocol: 'https'
}
v1client = VersiononeSdk::Client.new params

# Authorizing using Access Token Authentication
params = {
  hostname: 'www1.v1host.com',
  instance: 'myinstance',
  access_token: 'myaccesstoken',
  port: 443,
  protocol: 'https'
}
v1client = VersiononeSdk::Client.new params

# Different asset types can be defined and passed in during instantiation
params = {
  ...
  type_prefixes: { 'R' => 'Request', 'D' => 'Defect', 'B' => 'Story', 'E' => 'Epic' }
}
v1client = VersiononeSdk::Client.new params

# Retrieve an array of VersiononeSdk::Asset objects
assets   = v1client.getAssets('Scope')

assets.each do |asset|
  assetHash = asset.asHash
end

# Retrieve a single asset using an Asset OID Token
# Returns a VersiononeSdk::Asset object
asset = v1client.getAsset('Story:1')
asset = v1client.getAsset('Story',1)

# Retrieve a single asset using an Asset Number
asset = v1client.getAsset('B-1')
asset = v1client.getAsset('B',1)

# Updating an asset with a simple attribute
# Returns a Faraday::Response object
v1client.updateAsset('Member',20,'Phone','555-555-1212')
v1client.updateAsset('Member',20,'Phone',{value: '555-555-1212', act: 'set'})
v1client.updateAsset('Member',20,'Phone',{value: '555-555-1212', act: 'set'},\
  :simple_attribute
)

# Updating an asset with a single-value relationship:
v1client.updateAsset('Scope',0,'Owner','Member:20')
v1client.updateAsset('Scope',0,'Owner',{value: 'Member:20', act: 'set'})
v1client.updateAsset('Scope',0,'Owner',{value: 'Member:20', act: 'set'},:single_relationship)

# Updating an asset with a multi-value relationship: adding members
v1client.updateAsset('Scope',0,'Members',['Member:1000','Member:1001'],:multi_relationship)

# Updating an asset with a multi-value relationship: adding and removing members
v1client.updateAsset('Scope',0,'Members',[ \
  { value: 'Member:1000', act: 'add' },    \
  { value: 'Member:1001', act: 'remove' }  \
],:multi_relationship)
```

## Documentation

This gem is 100% documented with YARD, an exceptional documentation library. To see documentation for this, and all the gems installed on your system use:

```bash
$ gem install yard
$ yard server -g
```

## Notes

1. Integer Values
 - Integer values for Order and AssetState are converted from strings to integers.
2. Nil/Null Values and Empty Strings
 - The VersionOne JavaScript API provides empty strings when no value is present. This SDK uses Ruby nil values and JSON null values for empty strings. The primary reason for this is to support easy indexing using Elasticsearch.
3. Inflation
 - Some values are inflated. Currently, AssetState is used to derive AssetState.Name as defined here: https://community.versionone.com/Developers/Developer-Library/Concepts/Asset_State
4. Tracking Properties
 - In addition to the standard VersionOne properties, this modules adds the following generic properties for tracking: `:__id__sObjectDomain`, `:__id__sObjectType`, `:__id__iObjectId`, `:__id__sObjectUrl`. The object domain is set to 'Versionone', while object type and object id correspond to VersionOne Asset types and ids. The URL is the full URL for the resource including protocol, host and port.

## Change Log

See [CHANGELOG.md](CHANGELOG.md).

## Links

Project Repo

* https://github.com/grokify/versionone-sdk-ruby

VersionOne API Documentation

* http://community.versionone.com/Developers/Developer-Library/Documentation

VersionOne API Documentation for Updating an Asset

* https://community.versionone.com/Developers/Developer-Library/Recipes/Update_an_Asset

VersionOne Developer Google Group

* https://groups.google.com/forum/#!forum/versionone-dev

## Contributing

1. Fork it ( http://github.com/grokify/versionone-sdk-ruby/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Copyright and License

VersiononeSdk &copy; 2014-2016 by [John Wang](mailto:johncwang@gmail.com).

VersiononeSdk is licensed under the MIT license. Please see the [LICENSE.txt](LICENSE.txt) document for more information.

## Warranty

This software is provided "as is" and without any express or implied warranties, including, without limitation, the implied warranties of merchantibility and fitness for a particular purpose.

 [gem-version-svg]: https://badge.fury.io/rb/versionone_sdk.svg
 [gem-version-link]: http://badge.fury.io/rb/versionone_sdk
 [build-status-svg]: https://api.travis-ci.org/grokify/versionone-sdk-ruby.svg?branch=master
 [build-status-link]: https://travis-ci.org/grokify/versionone-sdk-ruby
 [coverage-status-svg]: https://coveralls.io/repos/grokify/versionone-sdk-ruby/badge.svg?branch=master
 [coverage-status-link]: https://coveralls.io/r/grokify/versionone-sdk-ruby?branch=master
 [dependency-status-svg]: https://gemnasium.com/grokify/versionone-sdk-ruby.svg
 [dependency-status-link]: https://gemnasium.com/grokify/versionone-sdk-ruby
 [codeclimate-status-svg]: https://codeclimate.com/github/grokify/versionone-sdk-ruby/badges/gpa.svg
 [codeclimate-status-link]: https://codeclimate.com/github/grokify/versionone-sdk-ruby
 [scrutinizer-status-svg]: https://scrutinizer-ci.com/g/grokify/versionone-sdk-ruby/badges/quality-score.png?b=master
 [scrutinizer-status-link]: https://scrutinizer-ci.com/g/grokify/versionone-sdk-ruby/?branch=master
 [downloads-svg]: http://ruby-gem-downloads-badge.herokuapp.com/versionone_sdk
 [downloads-link]: https://rubygems.org/gems/versionone_sdk
 [docs-rubydoc-svg]: https://img.shields.io/badge/docs-rubydoc-blue.svg
 [docs-rubydoc-link]: http://www.rubydoc.info/gems/versionone_sdk/
 [license-svg]: https://img.shields.io/badge/license-MIT-blue.svg
 [license-link]: https://github.com/grokify/versionone-sdk-ruby/blob/master/LICENSE.txt


CHANGELOG.md
CHANGELOG
---------
- **2016-06-10**: 0.2.1
  - Add raise asset not found error
  - Update `Gemfile` to use `gemspec`
- **2016-02-16**: 0.2.0
  - Add Access Token Authentication handling from pccasto. More info at: https://community.versionone.com/Developers/Developer-Library/Documentation/API_Authentication/Access_Token_Authentication
  - Update client's default protocol, hostname and port to use constants.
  - Fix Coveralls code coverage support.
- **2016-02-05**: 0.1.2
  - Update `VersiononeSdk::Asset` to update call to `JsonDoc::Document`
- **2015-09-14**: 0.1.1
  - Update `README.md`
  - Add Travis CI, Code Climate, Coveralls support
- **2014-03-24**: 0.1.0
  - Update JSON representation to use array for multiple values and string for single values for both attribute and relation XML tags. Previously relation tags would be converted to arrays even if there was only a single element. This change is not backward compatible.
  - Update control properties format from `__id__#{property}` to `_#{property}__id`. This change is not backward compatible.
- **2014-03-20**: 0.0.4
  - Fix for `VersiononeSdk::Update::updateAsset`
- **2014-03-20**: 0.0.3
  - Add ability to retrieve a single asset using an OID token or Number
  - Add `VersiononeSdk::Update` to support updating Assets
- **2014-03-17**: 0.0.2
  - Add `VersiononeSdk::Asset` object to support value inflation, starting with `AssetState.Name`
- **2014-03-16**: 0.0.1
  - Initial release


test/test_base.rb
require 'coveralls'
Coveralls.wear!

require 'test/unit'
require 'versionone_sdk'


test/test_setup.rb
require './test/test_base.rb'

class VersiononeSdkTest < Test::Unit::TestCase
  def testSetup
    oAsset  = VersiononeSdk::Asset.new
    oClient = VersiononeSdk::Client.new
    oParser = VersiononeSdk::ParserXmlAssets.new

    assert_equal 'VersiononeSdk::Asset', oAsset.class.name
    assert_equal 'VersiononeSdk::Client', oClient.class.name
    assert_equal 'VersiononeSdk::ParserXmlAssets', oParser.class.name
  end

  def testAsset
    oAsset = VersiononeSdk::Asset.new({hello: 'world'},{},false,false)
    assert_equal 'world', oAsset.getAttr(:hello)

    oAsset.setProp(:AssetState, '128')
    oAsset.inflateNames()

    assert_equal 'Closed', oAsset.getProp(:'AssetState.Name')
  end

  def testClient
    oClient = VersiononeSdk::Client.new(instance: 'Test')
    assert_equal 'https://localhost:443/Test/rest-1.v1/Data/',
                    oClient.getUrlForAssets
    assert oClient.oFaraday.ssl.verify

    oClient = VersiononeSdk::Client.new(instance: 'Test', ssl_verify: false)
    assert !(oClient.oFaraday.ssl.verify)

    oClient = VersiononeSdk::Client.new(protocol: 'http', user: 'user',
                password: 'pass', port: 80, access_token: '', instance: 'Test')
    assert_equal 'http://localhost/Test/rest-1.v1/Data/', oClient.getUrlForAssets
    assert_match(/\ABasic /, oClient.oFaraday.headers['Authorization'])

    sAccessToken = 'some_string_generated_from_version_one'
    oClient = VersiononeSdk::Client.new(protocol: 'https', port: 443,
                access_token: sAccessToken, instance: Test)
    assert_equal "Bearer #{sAccessToken}", oClient.oFaraday.headers['Authorization']

    oClient = VersiononeSdk::Client.new(instance: 'Test')
    assert_equal({'B' => 'Story', 'E' => 'Epic', 'D' => 'Defect'}, oClient.dTypePrefix)

    oClient = VersiononeSdk::Client.new(instance: 'Test', type_prefixes: { 'T' => 'Test' })
    assert_equal({ 'T' => 'Test'}, oClient.dTypePrefix)
  end
end


test/test_parser_xml_assets.rb
require './test/test_base.rb'

class ParserXmlAssetsTest < Test::Unit::TestCase
  def testGetJsondocForXmlAssetNodeNilNodeFailure
    oParser = VersiononeSdk::ParserXmlAssets.new

    assert_raise (RuntimeError.new("E_NIL_NODE")) do
      oParser.getJsondocForXmlAssetNode(nil)
    end
  end

  def testGetJsondocForXmlAssetNodeAssetNotFound
    oParser = VersiononeSdk::ParserXmlAssets.new
 
    assert_raise(RuntimeError.new("E_ASSET_NOT_FOUND /v1sdktesting/rest-1.v1/Data/Story/12345")) do
      errorDocument = Nokogiri::XML('<Error href="/v1sdktesting/rest-1.v1/Data/Story/12345"><Message>Not Found</Message></Error>')
      oParser.getJsondocForXmlAssetNode(errorDocument.child)
    end
  end
end


lib/versionone_sdk/asset.rb
require 'jsondoc'

module VersiononeSdk
  # VersiononeSdk::Asset class is a JsonDoc::Document subclass that includes
  # the #inflateNames method to add AssetState.Name based on the VersionOne
  # classifications avaiable at: https://community.versionone.com/Developers/Developer-Library/Concepts/Asset_State
  class Asset < JsonDoc::Document
    attr_accessor :dSchema

    def initialize(dValues=nil,dSchema=nil,bDefaultifyDoc=false,bIsStrict=true,opts={})
      opts[:bUseDeepKeys] = false
      super(dValues, dSchema, bDefaultifyDoc, bIsStrict, opts)
    end

    def getDefaultSchema()
      dSchema =  {}
      return dSchema
    end

    def inflate()
      self.inflateNames
      self.convertIntegers
    end

    def convertIntegers()
      [:Order,:AssetState].each do |yKey|
        xxVal = self.getProp(yKey)
        if xxVal.is_a?(String) && xxVal =~ /^-?[0-9]+$/
          xxVal = xxVal.to_i
          self.setProp(yKey,xxVal)
        end
      end
    end

    def inflateNames()
      dAssetState = {
        0   => 'Future',
        64  => 'Active',
        128 => 'Closed',
        200 => 'Template (Dead)',
        208 => 'Broken Down (Dead)',
        255 => 'Deleted (Dead)'
      }
      if @dDocument.key?(:AssetState)
        sAssetState = @dDocument[:AssetState]
        if sAssetState.is_a?(String) && sAssetState =~ /^[0-9]+$/
          iAssetState = sAssetState.to_i
          if dAssetState.key?(iAssetState)
            sAssetStateName = dAssetState[iAssetState]
            self.setProp(:'AssetState.Name',sAssetStateName)
          end
        end
      end
    end
  end
end


lib/versionone_sdk/update.rb
require 'nokogiri'

module VersiononeSdk
  class Update
    def initialize(oClient=nil)
      @oClient = oClient
      @dTagTypes = {simple_attribute: 1, single_relationship: 1, multi_relationship: 1}
      @sRelationships = 'BuildProjects,Owner,Parent,Schedule,Scheme,SecurityScope,Status,TestSuite'
      @dRelationships = Hash[@sRelationships.split(',').collect {|v| [v,1]}]
    end

    # Update a VersionOne asset.
    #
    # Params:
    # +sAssetType+:: A REST API asset type such as Scope, Epic, Story, Member, etc. The first part of the OID Token.
    # +sAssetOid+:: The numerical OID and the second part of the OID token.
    # +sName+:: Name of attribute to be updated.
    # +xxValues+:: values to be updated which can be a variety of formats.
    # +yTagType+:: A optional symbol to identify the type of attribute, e.g.
    # :simple_attribute, :single_relationship, or :multi_relationship
    def updateAsset(sAssetType=nil,sAssetOid=nil,sName=nil,xxValues=nil,yTagType=nil)
      aValues  = normalizeValues(xxValues)
      # Validate Tag Type
      yTagType = yTagType.to_sym if yTagType.is_a?(String)

      unless yTagType.nil?
        unless @dTagTypes.key?(yTagType)
          raise ArgumentError, "E_BAD_TAG_TYPE: [#{yTagType.to_s}]"
        end
      else
        if sName.nil? || ! sName.kind_of?(String)
          raise ArgumentError, 'E_NO_ATTRIBUTE_NAME'
        elsif @dRelationships.key?(sName)
          aValues.each do |dValue|
            sAct = dValue[:act]
            if sAct    == 'set'
              yTagType =  :single_relationship
            elsif sAct == 'add' || sAct == 'remove'
              yTagType =  :multi_relationship
            else
              raise ArgumentError, "E_BAD_ACT: [#{sAct}]"
            end
          end
        else
          yTagType = :simple_attribute
        end
      end

      xBody \
        = yTagType == :simple_attribute    \
        ? getXmlBodyForSimpleAttributeUpdate(sName,aValues) \
        : yTagType == :single_relationship \
        ? getXmlBodyForSingleRelationship(sName,aValues)    \
        : yTagType == :multi_relationship  \
        ? getXmlBodyForMultiRelationship(sName,aValues)     \
        : nil

      unless xBody.nil?
        oFdRes = @oClient.oFaraday.post do |req|
          req.url getUrlForAssetTypeAndOid(sAssetType,sAssetOid)
          req.headers['Content-Type'] = 'application/xml'
          req.body = xBody
        end
        return oFdRes
      end

      return nil

    end

    private

    def normalizeValues(xxValues=nil)
      aValues = []

      if xxValues.is_a?(String)
        aValues = [{value: xxValues, act: 'set'}]
      elsif xxValues.is_a?(Hash)
        aValues = [xxValues]
      elsif xxValues.is_a?(Array)
        xxValues.each do |xxSubValue|
          if xxSubValue.is_a?(String)
            sAct = xxValues.length > 1 ? 'add' : 'set'
            aValues.push({value: xxSubValue, act: sAct})
          elsif xxSubValue.is_a?(Hash)
            aValues.push(xxSubValue)
          end
        end
      end

      return aValues
    end

    def getUrlForAssetTypeAndOid(sAssetType=nil,sAssetOid=nil)
      sUrl = File.join('/'+@oClient.sInstance,'rest-1.oauth.v1/Data',sAssetType,sAssetOid.to_s)
      return sUrl
    end

    def getXmlBodyForSimpleAttributeUpdate(sName=nil,aValues=[])
      sValue   = aValues.length>0 && aValues[0].key?(:value) \
        ? aValues[0][:value] : nil
      oBuilder = Nokogiri::XML::Builder.new do |xml|
        xml.Asset {
          xml.Attribute(sValue, name: sName, act: 'set')
        }
      end
      return oBuilder.to_xml
    end

    def getXmlBodyForSingleRelationship(sName=nil,aValues=[])
      sValue   = aValues.length>0 && aValues[0].key?(:value) \
        ? aValues[0][:value] : nil
      oBuilder = sValue.nil? \
        ? Nokogiri::XML::Builder.new { |xml| xml.Asset { \
            xml.Relation(name: sName, act: 'set')
        } }
        : Nokogiri::XML::Builder.new { |xml| xml.Asset { \
            xml.Relation(name: sName, act: 'set') {
              xml.Asset(idref: sValue)
            }
        } }
      return oBuilder.to_xml
    end

    def getXmlBodyForMultiRelationship(sName=nil,aValues=[])
      oBuilder = aValues.length == 0 \
        ? Nokogiri::XML::Builder.new { |xml| xml.Asset { \
            xml.Relation(name: sName, act: 'set')
        } }
        : Nokogiri::XML::Builder.new { |xml| xml.Asset { \
            xml.Relation(name: sName) {
              aValues.each do |dValue|
                xml.Asset(idref: dValue[:value], act: dValue[:act])
              end
            }
        } }
      return oBuilder.to_xml
    end
  end
end


lib/versionone_sdk/parser_xml_assets.rb
require 'jsondoc'
require 'nokogiri'
require 'versionone_sdk/asset'

module VersiononeSdk
  class ParserXmlAssets
    def initialize(dOptions={})
      @sUrl = dOptions.key?(:url) ? dOptions[:url] : ''
    end
    def getDocsForAssetsXmlPath(sPath=nil)
      if File.exists?(sPath)
        return self.getDocsForAssetsXml(IO.read(sPath))
      end
      return []
    end
    def getDocsForAssetsXml(xAssets=nil)
      oXml = Nokogiri::XML::Document.parse(xAssets)
      oBlock = oXml.xpath("//Assets/Asset")
      aDocs = []
      oBlock.map do |oNodeAsset|
        oAsset = self.getJsondocForXmlAssetNode( oNodeAsset )
        unless oAsset.nil?
          aDocs.push(oAsset)
        end
      end
      return aDocs
    end
    def getDocForAssetXml(xAsset=nil)
      oXml = Nokogiri::XML::Document.parse(xAsset)
      oAsset = self.getJsondocForXmlAssetNode( oXml.root )
      return oAsset
    end
    def getJsondocForXmlAssetNode(oNodeAsset=nil)
      if oNodeAsset.nil?
        raise RuntimeError, 'E_NIL_NODE'
      end
      oAsset = VersiononeSdk::Asset.new
      oAsset.bIsStrict = false
      sOidtokSrc       = nil
      sObjectType      = nil
      iObjectId        = nil
      if oNodeAsset.attribute('id')
        sOidtokSrc     =  oNodeAsset.attribute('id').value
        if sOidtokSrc  =~ /^(.+):([0-9]+)$/
          sObjectType  =  $1
          iObjectId    =  $2.to_i
          oAsset.setProp(:_sObjectDomain__id,'Versionone')
          oAsset.setProp(:_sObjectType__id,sObjectType)
          oAsset.setProp(:_iObjectId__id,  iObjectId)
        end
      end
      if oNodeAsset.attribute('href').value
        sUrl = @sUrl \
          ? @sUrl + oNodeAsset.attribute('href').value
          : oNodeAsset.attribute('href').value
        oAsset.setProp(:_sObjectUrl__id,sUrl)
      end
      oNodeAsset.children.each do |oNodeChild|
        if oNodeChild.name == 'Attribute' || oNodeChild.name == 'Relation'
          if oNodeChild.attribute('name')
            yPropKey = oNodeChild.attribute('name').to_s.to_sym
            xxPropVal = getAssetNodeChildPropVal( oNodeChild )
            oAsset.setProp(yPropKey,xxPropVal)
          end
        elsif oNodeChild.name == 'Message' && oNodeChild.text == 'Not Found'
          raise RuntimeError, "E_ASSET_NOT_FOUND #{oNodeAsset.attribute('href').value}"
        else
          raise RuntimeError, "E_UNKNOWN_ASSET_NODE_NAME #{oNodeChild.name}"
        end
      end
      oAsset.inflate
      return oAsset
    end

    def getAssetNodeChildPropVal(oNodeChild=nil)
      xxPropVal = nil
      aPropVals = []
      if oNodeChild.children.length > 0
        oNodeChild.children.each do |oNodeChildGrand|
          dAttributes = oNodeChildGrand.attributes
          xxPropVal   = dAttributes.key?('idref')    \
            ? oNodeChildGrand.attribute('idref').value \
            : oNodeChildGrand.text
          xxPropVal = nil if xxPropVal == ''
          aPropVals.push(xxPropVal)
        end
        if aPropVals.length > 1
          xxPropVal = aPropVals
        elsif aPropVals.length == 1
          xxPropVal == aPropVals[0]
        else
          xxPropVal = nil
        end
      else
        xxPropVal = oNodeChild.text
      end
      xxPropVal = nil if xxPropVal == ''
      return xxPropVal
    end
  end
end


lib/versionone_sdk/client.rb
require 'faraday'
require 'versionone_sdk/update'
require 'versionone_sdk/parser_xml_assets'

module VersiononeSdk
  class Client
    DEFAULT_PROTOCOL = 'https'
    DEFAULT_HOSTNAME = 'localhost'
    DEFAULT_PORT = 443

    attr_accessor :oFaraday
    attr_accessor :sInstance
    attr_accessor :dTypePrefix

    def initialize(opts = {})
      @sProtocol   = opts[:protocol] || DEFAULT_PROTOCOL
      @sHostname   = opts[:hostname] || DEFAULT_HOSTNAME
      @iPort       = opts.key?(:port) && opts[:port] \
                   ? opts[:port].to_i : DEFAULT_PORT
      sUsername    = opts[:username] || ''
      sPassword    = opts[:password] || ''
      # VersionOne provides a mechanism for generating an access token
      sAccessToken = opts[:access_token] || ''
      @sInstance   = opts[:instance] || ''
      @dTypePrefix = opts[:type_prefixes] || {'B' => 'Story', 'E' => 'Epic', 'D' => 'Defect'}
      @sUrl        = buildUrl(@sProtocol, @sHostname, @iPort)
      @oFaraday    = Faraday::Connection.new url: @sUrl
      @oFaraday.ssl.verify = opts[:ssl_verify].to_s.match(/false/i) \
                   ? false : true
      if sAccessToken.empty?
        @oFaraday.basic_auth(sUsername, sPassword)
      else
        # could also patch Faraday to have a method similar to basic_auth
        @oFaraday.headers['Authorization']  = "Bearer #{sAccessToken}"
      end
      @oUpdate     = VersiononeSdk::Update.new self
    end

    def getAsset(xAssetId1, xAssetId2 = nil)
      xAssetId1.strip!

      if xAssetId1 =~ /^([^:]+):([0-9]+)$/
        sAssetType = $1
        sAssetOid = $2.to_i
        return self.getAssetForTypeAndOid(sAssetType, sAssetOid)

      elsif xAssetId1 =~ /^([a-zA-Z])-[0-9]+$/
        sAssetTypeAbbr = $1.upcase
        sAssetType = @dTypePrefix.key?(sAssetTypeAbbr) \
          ? @dTypePrefix[ sAssetTypeAbbr ] : ''
        xAssetId1.upcase!
        return self.getAssetForTypeAndNumber(sAssetType, xAssetId1)

      elsif !xAssetId2.nil?
        if xAssetId2.is_a?(String) && xAssetId2 =~ /^[0-9]+$/
          xAssetId2 = xAssetId2.to_i
        end

        if xAssetId2.is_a?(Integer)

          if xAssetId1 =~ /^[a-zA-Z]$/
            xAssetId1.upcase!
            sAssetTypeAbbr = xAssetId1
            sAssetType = @dTypePrefix.key?(sAssetTypeAbbr) \
              ? @dTypePrefix[ sAssetTypeAbbr ] : ''
            sAssetNumber = xAssetId1 + '-' + xAssetId2.to_s
            sAssetNumber.upcase!
            return self.getAssetForTypeAndNumber(sAssetType, sAssetNumber)
          elsif xAssetId1 =~ /^[a-zA-Z].+$/
            return self.getAssetForTypeAndOid(xAssetId1, xAssetId2)
          end
        end
      end

      raise RuntimeError, "E_UNKNOWN_ASSET_ID [#{xAssetId1}][#{xAssetId2.to_s}]"
    end

    def getAssetForTypeAndOid(sAssetType = nil, sAssetOid = nil)
      sUrl = self.getUrlForAssets( sAssetType, sAssetOid )
      oRes = @oFaraday.get sUrl
      oParser = VersiononeSdk::ParserXmlAssets.new({url: @sUrl})
      return oParser.getDocForAssetXml( oRes.body )
    end

    def getAssetForTypeAndNumber(sAssetType = nil, sAssetNumber = nil)
      sUrl = self.getUrlForAssetTypeAndNumber( sAssetType, sAssetNumber )
      oRes = @oFaraday.get sUrl
      oParser = VersiononeSdk::ParserXmlAssets.new({url: @sUrl})
      aDocs = oParser.getDocsForAssetsXml( oRes.body )
      return aDocs[0]
    end

    def getAssets(sAssetType = nil, xIds = nil)
      oRes = self.getAssetsXml(sAssetType,xIds)
      oParser = VersiononeSdk::ParserXmlAssets.new({url: @sUrl})
      return oParser.getDocsForAssetsXml( oRes.body )
    end

    def getAssetsXml(sAssetType = nil, xIds = nil)
      sUrl = self.getUrlForAssets(sAssetType)
      return @oFaraday.get sUrl
    end

    def getUrlForAssetTypeAndNumber(sAssetType = nil, sAssetNumber = nil)
      aUrl = [ @sUrl, @sInstance, 'rest-1.v1/Data',sAssetType + %Q!?where=Number="#{sAssetNumber}"!]
      return aUrl.join('/')
    end

    def getUrlForAssets(sAssetType = nil, sAssetOid = nil)
      aUrl = [@sUrl, @sInstance, 'rest-1.v1/Data',sAssetType]
      if sAssetOid.is_a?(Integer)
        aUrl.push sAssetOid
      elsif sAssetOid.kind_of?(String) && sAssetOid =~ /^[0-9]+$/
        aUrl.push sAssetOid
      end
      return aUrl.join('/')
    end

    def updateAsset(sAssetType=nil,sAssetOid=nil,sName=nil,xxValues=nil,yTagType=nil)
      return @oUpdate.updateAsset(sAssetType, sAssetOid, sName, xxValues, yTagType)
    end

    private

    def buildUrl(sProtocol = DEFAULT_PROTOCOL, sHostname = DEFAULT_HOSTNAME, iPort = DEFAULT_PORT)
      if sHostname.nil?
        sHostname = 'localhost'
      elsif sHostname.is_a?(String)
        sHostname.strip!
        if sHostname.length < 1
          sHostname = 'localhost'
        end
      else
        raise ArgumentError, 'E_HOSTNAME_IS_NOT_A_STRING'
      end
      if iPort.nil?
        iPort = 80
      elsif iPort.is_a?(String) && iPort =~ /^[0-9]+$/
        iPort = iPort.to_i
      elsif ! iPort.kind_of?(Integer)
        raise ArgumentError, 'E_PORT_IS_NOT_AN_INTEGER'
      end
      sBaseUrl = "#{sProtocol}://#{sHostname}"
      sBaseUrl.sub!(/\/+\s*$/,'')
      sBaseUrl += ':' + iPort.to_s if iPort != 80
      return sBaseUrl
    end
  end
end
